src/crypto.rs

```
@@ -1,196 +1,204 @@
-            description("Decrypt operation is not allowed within this crypto context")
+    }
+}
+
+quick_error! {
+    #[derive(Clone, Debug)]
+    pub enum DecryptBytesError {
+        DecryptVerify {
+            description("error decrypting/verifying message")
         }
     }
 }
 
-/// Simplifies encryption/decryption by holding the necessary context - encryption keys.
-/// Allows "null" encryption where data is only serialized. See: null object pattern.
-#[derive(Clone, Debug)]
-pub enum CryptoContext {
-    /// No encryption.
-    Null,
-    /// Encryption + authentication
-    Authenticated {
-        /// Their public key.
-        their_pk: PublicKey,
-        /// Our secret key.
-        our_sk: SecretKey,
-    },
-    /// No message authentication. Only decrypt operation is allowed.
-    AnonymousDecrypt {
-        /// Our private key.
-        our_pk: PublicKey,
-        /// Our secret key.
-        our_sk: SecretKey,
-    },
-    /// No message authentication. Only encrypt operation is allowed.
-    AnonymousEncrypt {
-        /// Their public key.
-        their_pk: PublicKey,
-    },
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Clone)]
+pub struct P2pPublicId {
+    sign: sign::PublicKey,
+    encrypt: box_::PublicKey,
 }
 
-impl CryptoContext {
-    /// Construct crypto context that encrypts and authenticate messages.
-    pub fn authenticated(their_pk: PublicKey, our_sk: SecretKey) -> Self {
-        CryptoContext::Authenticated { their_pk, our_sk }
-    }
+impl PublicId for P2pPublicId {
+    type Signature = sign::Signature;
 
-    /// Contructs "null" encryption context which actually does no encryption.
-    /// In this case data is simply serialized but not encrypted.
-    pub fn null() -> Self {
-        CryptoContext::Null
+    fn encrypt_anonymous<T>(&self, plaintext: &T) -> Vec<u8>
+    where
+        T: Serialize,
+    {
+        let bytes = unwrap!(serialisation::serialise(plaintext));
```

I know we had this discussion, but I think we should propagate this error.  `serialise` boils down to https://docs.rs/bincode/1.0.0/bincode/fn.serialize.html which returns `Result` itself. It's not clear how serialisation can fail, but if underlying library says it does, I don't think we're safe unwrapping.

---

It can fail because it's possible to use a size limit when calling that function and also because the `serde::Serialize` trait allows you to return an error (though we never make use of this). Given that encryption really can never fail, it would be nice to not have to have an `Encrypt` error.

One pattern I'd like to experiment with is having error enum variants that only exist for release. Similar to how `debug_assert!` panics on debug mode and does nothing on release mode, or how integer overflow panics on debug mode and wraps on release mode, it would be nice to have error variants that panic on creation in debug mode but are propagated normally in release mode. This may be more effort than it's worth though.